presto-spi/src/main/java/com/facebook/presto/spi/block/DictionaryBlockEncoding.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.block;
+
+import com.facebook.presto.spi.type.TypeManager;
+import io.airlift.slice.Slice;
+import io.airlift.slice.SliceInput;
+import io.airlift.slice.SliceOutput;
+
+import static java.util.Objects.requireNonNull;
+
+public class DictionaryBlockEncoding
+        implements BlockEncoding
+{
+    public static final BlockEncodingFactory<DictionaryBlockEncoding> FACTORY = new DictionaryBlockEncodingFactory();
+    private static final String NAME = "DICTIONARY";
+    private final BlockEncoding dictionaryEncoding;
+
+    public DictionaryBlockEncoding(BlockEncoding dictionaryEncoding)
+    {
+        this.dictionaryEncoding = requireNonNull(dictionaryEncoding, "dictionaryEncoding is null");
+    }
+
+    @Override
+    public String getName()
+    {
+        return NAME;
+    }
+
+    @Override
+    public void writeBlock(SliceOutput sliceOutput, Block block)
+    {
+        // The down casts here are safe because it is the block itself the provides this encoding implementation.
+        DictionaryBlock dictionaryBlock = (DictionaryBlock) block;
+
+        try {
+            dictionaryBlock = dictionaryBlock.compact();
+        }
+        catch (UnsupportedOperationException e) {
+            // ignore if compacting a dictionary is not supported
```

Comment:
I should handle this in the ``compact`` code and return the same dictionary. I'll change it.